MHArmory.Search/OpenCL/SearchDataSerializer.cs

```diff
@@ -0,0 +1,243 @@
+        {
+            var parameters = new SerializedSearchParameters();
+            
+            var heads = data.AllHeads.Where(x => x.IsSelected).Select(x => x.Equipment).ToList();
+            var chests = data.AllChests.Where(x => x.IsSelected).Select(x => x.Equipment).ToList();
+            var gloves = data.AllGloves.Where(x => x.IsSelected).Select(x => x.Equipment).ToList();
+            var waists = data.AllWaists.Where(x => x.IsSelected).Select(x => x.Equipment).ToList();
+            var legs = data.AllLegs.Where(x => x.IsSelected).Select(x => x.Equipment).ToList();
+            var charms = data.AllCharms.Where(x => x.IsSelected).Select(x => x.Equipment).ToList();
+
+            var allArmorPieces = heads.Concat(chests).Concat(gloves).Concat(waists).Concat(legs).Cast<IArmorPiece>().ToList();
+            var allEquipment = allArmorPieces.Concat(charms).ToList();
+
+            var armorSetParts = allArmorPieces.Where(x => x.ArmorSetSkills != null).SelectMany(x => x.ArmorSetSkills.SelectMany(y => y.Parts)).ToList();
+            var desiredAbilitiesSkillSet = new HashSet<int>(data.DesiredAbilities.Select(x => x.Skill.Id));
+            IEnumerable<IArmorSetSkillPart> desiredArmorSetParts = armorSetParts.Where(x => x.GrantedSkills.Any(y => desiredAbilitiesSkillSet.Contains(y.Skill.Id)));
+            IEnumerable<int> desiredSkillIDs = data.DesiredAbilities.Select(x => x.Skill.Id);
+
+            var maps = new SearchIDMaps();
+
+            maps.EquipmentIdMap = CreateIDMap(allEquipment.Select(x => new Tuple<int, EquipmentType>(x.Id, x.Type)));
+            if (maps.EquipmentIdMap.Count > SearchLimits.TotalEquipments)
+            {
+                throw new OpenCLSerializationException($"Tried to serialize {maps.EquipmentIdMap.Count} equipment pieces, maximum allowed is {SearchLimits.TotalEquipments}");
+            }
+
+            maps.JewelIdMap = CreateIDMap(data.AllJewels.Select(x => x.Jewel.Id));
+            if (maps.EquipmentIdMap.Count > SearchLimits.TotalJewels)
+            {
+                throw new OpenCLSerializationException($"Tried to serialize {maps.JewelIdMap.Count} distinct jewels, maximum allowed is {SearchLimits.TotalJewels}");
+            }
+
+            maps.SetIdMap = CreateIDMap(desiredArmorSetParts.Select(x => x.Id));
+            if (maps.SetIdMap.Count > SearchLimits.MaxSetSkills)
+            {
+                throw new OpenCLSerializationException($"Tried to serialize {maps.SetIdMap.Count} distinct set parts, maximum allowed is {SearchLimits.MaxSetSkills}");
+            }
+
+            maps.SkillIdMap = CreateIDMap(desiredSkillIDs);
+            if (maps.SkillIdMap.Count > SearchLimits.MaxDesiredSkills)
+            {
+                throw new OpenCLSerializationException($"Tried to serialize {maps.SkillIdMap.Count} distinct skills, maximum allowed is {SearchLimits.MaxDesiredSkills}");
+            }
+
+            byte[] header = new byte[13];
```

Comment:
Maybe that would be better to use a `List<byte>` and add to it, and then in the end call `.ToArray()` if need be to make it an array. In this case, not need for hard-coded values (such as the 13) and index all items with hard coded values as well.

Reply:
I think it's even better to use an expanding `MemoryStream` and a `BinaryWriter` here, the header first had just two values, then it grew, and grew, and grew, and I never got to actually properly rewriting it.

The reason that a `BinaryWriter` would be even better is that if someone ever needs to write a value which isn't a byte, the `BinaryWriter` is a lot more convenient.